### PR TITLE
Add auto-improvement classification blog draft

### PR DESCRIPTION
**Pull request by annabellscha**
## Summary
- add the draft blog post for the auto-improvement classification experiment
- include the loop diagram screenshot and the Fable 5 loops screenshot
- preserve the latest intro and model-cost framing from the preview edits

### CODE REVIEW

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 1]
```diff
@@ -0,0 +1,108 @@
+---
+title: "97% on train, 82% on test: auto-improvement loops have a validation gap"
+date: 2026/06/10
+description: "Claude Fable 5 had just come out, so we used it to run an auto-improvement loop end to end. It improved train accuracy fast. What it really taught us was how much a validation split matters."
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: the word "matters" has become an AI word for me

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: could say "the importance of a good validation split"

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+tag: engineering
+author: Annabell
+---
+
+import { BlogHeader } from "@/components/blog/BlogHeader";
+import { Frame } from "@/components/Frame";
+
+<BlogHeader
+  title="97% on train, 82% on test: auto-improvement loops have a validation gap"
+  description="Claude Fable 5 had just come out, so we used it to run an auto-improvement loop end to end. It improved train accuracy fast. What it really taught us was how much a validation split matters."
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I would address that steipete is talking about personal coding agents, and Langfuse generally is used for agents that you deploy for users

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  authors={["annabellschafer"]}
+/>
+
+Loop-shaped workflows are having a moment. `@steipete` put it plainly: you should not be prompting coding agents anymore, you should be designing loops that prompt your agents.
+
+<Frame>
+  ![Peter Steinberger post about designing loops that prompt your agents](/images/blog/2026-06-10-auto-improvement-classification/peter-steinberger-design-loops.png)
+</Frame>
+
+We designed a loop and gave Claude Fable 5 a classification task: a train/test split in [Langfuse Datasets](/docs/evaluation/experiments/datasets), a prompt in [Prompt Management](/docs/prompt-management/get-started), and a goal: hit 95% accuracy or stop at 15 runs. Train accuracy went from 78% to 97% in four runs. Test performance barely moved. The 11 remaining errors on test were shared across every prompt variant.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I'd not mention too many specifics here, and instead mention this in the "the task and setup" section. This paragraph could become

We designed a loop and gave Claude Fable 5 a classification task. The loop did its job, but it surfaced a dataset problem: while train accuracy improved, test performance barely moved.

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+The loop did its job. What it surfaced was a dataset problem.
+
+## The task and setup
+
+Our task was to classify arXiv papers into one of 10 categories from title, authors, and abstract, using this [Kaggle dataset](https://www.kaggle.com/datasets/sumitm004/arxiv-scientific-research-papers-dataset). We picked classification because it gives you a crisp target function: exact-match accuracy.
+
+- a train split with 200 labeled examples and a held-out test split with 100 in [**Langfuse Datasets**](/docs/evaluation/experiments/datasets)
+- a prompt in [**Prompt Management**](/docs/prompt-management/get-started)
+- a small runner built with [**Langfuse Experiments via the SDK**](/docs/evaluation/experiments/experiments-via-sdk)
+- `gpt-4o-mini` as the task model<sup id="fnref-model-choice"><a href="#fn-model-choice">1</a></sup>
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I'd explain the improvement loop before this, and explicitly mention that there is a task model, used for the classification task, and an optimizer model, that runs experiments to improve that classification task. If not explained people might be confused because you mentioned using Claude Fable 5 in the intro, and here it mentions gpt-4o-mini

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: also, the o renders weird, not sure why though? 

Image

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+The starting prompt was minimal, consisting only of a list of labels with instructions to pick one. After each run, the agent reviewed the errors, wrote [comments](/docs/observability/features/comments) on the [Langfuse trace](/docs/observability/data-model#traces), published a new [prompt version](/docs/prompt-management/features/prompt-version-control), and ran again.
+
+_Suggested platform screenshot: the Langfuse workbench for this loop, showing the dataset, prompt, and experiment setup together._
+
+<Frame fullWidth>
+  ![Diagram of the autonomous training loop: start from a base prompt, run the train split, score rows, comment on errors, revise the prompt, and then run the held-out test split once](/images/blog/2026-06-10-auto-improvement-classification/fable-classification-loop-diagram.png)
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:47]: would make the formatting better here 

- "revise and run again" overlaps with TRAIN LOOP box edge
- "pass ->" too close to the revise prompt box

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:47]: I'd also try to remove the bottom line of text in each box. It feels a bit noisy now

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+## The task and setup
+
+Our task was to classify arXiv papers into one of 10 categories from title, authors, and abstract, using this [Kaggle dataset](https://www.kaggle.com/datasets/sumitm004/arxiv-scientific-research-papers-dataset). We picked classification because it gives you a crisp target function: exact-match accuracy.
+
+- a train split with 200 labeled examples and a held-out test split with 100 in [**Langfuse Datasets**](/docs/evaluation/experiments/datasets)
+- a prompt in [**Prompt Management**](/docs/prompt-management/get-started)
+- a small runner built with [**Langfuse Experiments via the SDK**](/docs/evaluation/experiments/experiments-via-sdk)
+- `gpt-4o-mini` as the task model<sup id="fnref-model-choice"><a href="#fn-model-choice">1</a></sup>
+
+The starting prompt was minimal, consisting only of a list of labels with instructions to pick one. After each run, the agent reviewed the errors, wrote [comments](/docs/observability/features/comments) on the [Langfuse trace](/docs/observability/data-model#traces), published a new [prompt version](/docs/prompt-management/features/prompt-version-control), and ran again.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:42]: Would be nice to have the first prompt version shown, could be in an expandable component if it takes up much space

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+_Suggested platform screenshot: the Langfuse workbench for this loop, showing the dataset, prompt, and experiment setup together._
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:44]: Just flagging, I think this is still a todo in the doc

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:44]: (I like to mark with yellow, makes it impossible to look over :) )

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+| Prompt | Train | Test | Gap |
+| --- | ---: | ---: | ---: |
+| v1 - flat label list | 78.0% | - | - |
+| v2 - general definitions | 90.5% | **84.0%** | 6.5 |
+| v3 - sharpened boundary rules | 90.0% | - | - |
+| v4 - train-derived precedents | **97.0%** | 82.0% | **15.0** |
+
+Moving from a flat label list to general definitions was real progress. But once the agent started encoding concrete precedents from the training failures, train accuracy jumped while generalization got worse. The prompt that looked best on train was not the one that held up on test.
+
+_Suggested platform screenshot: the Langfuse runs overview comparing prompt versions and scores across the experiment history._
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx:63]: also flagging this todo

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+The first round optimized fast:
+
+| Prompt | Train | Test | Gap |
+| --- | ---: | ---: | ---: |
+| v1 - flat label list | 78.0% | - | - |
+| v2 - general definitions | 90.5% | **84.0%** | 6.5 |
+| v3 - sharpened boundary rules | 90.0% | - | - |
+| v4 - train-derived precedents | **97.0%** | 82.0% | **15.0** |
+
+Moving from a flat label list to general definitions was real progress. But once the agent started encoding concrete precedents from the training failures, train accuracy jumped while generalization got worse. The prompt that looked best on train was not the one that held up on test.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: "was real progress" sounds a bit weird

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+_Suggested platform screenshot: the Langfuse runs overview comparing prompt versions and scores across the experiment history._
+
+## Round 2: "generalize this time"
+
+So we restarted from the more general prompt and changed the rules: no single-paper precedents, only class-level principles, and no touching the test set.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: "no touching the test set": I assume they also couldn't touch the test set before? If so, remove this part. 
If they could touch the test set before, this needs to be clarified (and defended)

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+| Prompt | Train | Test | Gap |
+| --- | ---: | ---: | ---: |
+| v2 - general definitions, round 1 | 90.5% | **84.0%** | 6.5 |
+| v5 - reasoning field, round 2 | 84.0% | - | - |
+| v9 - general principles, round 2 | 94.0% | 81.0% | 13.0 |
+
+Only selected prompts were run on the held-out test split.
+
+The disciplined second round did not produce better held-out results. Adding a `reasoning` field also did not help: in our runs, it seemed to encourage the model to rationalize surface cues instead of resolving the label boundary.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: Here I'm a bit lost
- "disciplined round"?
- rationalize surface cues instead of resolving label boundary: what does that mean? --> best to give an example, or leave it out

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+By the end, Fable's own analysis suggested that many of the remaining errors sat on genuinely fuzzy category boundaries, so it stopped before hitting the 95% target. Pushing further likely would have required adding increasingly specific case-by-case rules, rather than finding broader principles that generalized.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: genuinely = AI tell

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+The final prompt version of round 2 hit 81.0% accuracy on test. That is a 13-point gap to the train set, and very much in line with the drop between train and test in round 1. It just took more turns to get to the point of overfitting.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I was confused with round 1 / round 2 and the iteration numbers. Can we use something other than numbers for the rounds?

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+In the end, the best-performing prompt on test was still `v2` from the first round. That iteration simply added broad descriptions to the label-only version, and it generalized best.
+
+_Suggested platform screenshot: one recurring hard example in Langfuse, with the trace and annotation showing why the boundary case stayed unresolved._
+
+## What the dataset would have needed
+
+Based on these runs, the dataset likely needed three things:
+
+**1. A real validation split.** We had train for fitting and test for the final check, but nothing in between. So the loop selected prompt versions on train accuracy alone.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: This is not clear to me, does this mean we needed a third set? What would the function of that set be?

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+**2. More repeated edge cases.** The hard errors clustered around a few label boundaries, especially Information Retrieval vs. Databases, Human-Computer Interaction vs. Computers and Society, and subject vs. representation for audio papers. A stronger benchmark would have forced new rules to prove themselves across multiple similar cases.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: "The hard errors clustered around a few label boundaries, especially Information Retrieval vs. Databases, Human-Computer Interaction vs. Computers and Society, and subject vs. representation for audio papers." --> The section feels like a conclusion section to me, I don't expect to read new information here.  I'd discuss these specific errors earlier and then reference them here

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: benchmark --> dataset (we call it dataset earlier)

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: "forced **new rules** to prove themselves across **multiple similar cases.**" --> this sounds contradictory, what exactly does this mean?

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+**3. Clearer policy for ambiguous papers.** Some of the shared errors look genuinely arguable. If the benchmark wants one exact label, it needs sharper tie-break rules, better canonical examples, and maybe even an `unsure` or multi-label policy.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: genuinely = AI tell

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I feel like this point and point 2 are about the same thing, maybe cut one of them?

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: benchmark -> dataset

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+That is the real lesson: the loop did its job. It surfaced, quickly, that the next bottleneck was not another prompt tweak. It was the dataset.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: I'd leave this out

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+## Where this is actually useful today
+
+None of this means "do not automate the loop." It means: automate the inner loop, own the outer one.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: "automate the inner loop, not the outer one": what do you mean by that? In my head the 2 loops are 
1.  the auto-improvement loop that Fable runs (outer)
2. classification loop on the dataset by gpt 40 (inner)
(but that wouldn't make sense given this post is about auto improvement)

Comment (Lotte-Verheyden) [reply content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: Ah reading further I see what you mean with each. I wonder if the human-owned part is actually a loop though (the post also didn't describe it as a loop), would actually cut the inner vs outer loop here, I think it would confuse people

File: content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+- **Agent-owned:** running experiments, scoring, per-error annotation, drafting hypothesis-driven prompt revisions, diffing errors across runs, flagging plateaus
+- **Human-owned:** the target function, including the validation and held-out test data nobody optimizes against, dataset composition, when to restart with different constraints, and when to stop
+
+As we argued in [AI is eating the AI engineering loop](/blog/2026-06-09-ai-is-eating-ai-engineering), the mechanics aren't the hard part anymore. This experiment shows what the hard part actually is: the target function, the dataset, and the judgment calls nobody automates against.
```
Comment (Lotte-Verheyden) [inline content/blog/2026-06-10-what-we-learned-running-auto-improvement-for-classification.mdx]: Reading this, I would actually frame this post as an illustration of how lack of human judgement  in parts of the AI engineering loop result in suboptimal agents.

**Review by Lotte-Verheyden — COMMENTED** [review]
I mainly left comments on writing style and detailed content, but in general I think the story/framing could be stronger. 
Currently it reads as "we tried this setup, but didn't put enough time in looking at the dataset before starting, but we still want to discuss the results,  even though they aren't very good (because of the dataset)". My main thought as a reader would be "you should have done it again with a better dataset before writing about it". 
--> there is still a lot of interesting stuff in here, but because the story/framing isn't there, it doesn't feel intentional, and it's not clear what the reader should take from this. 

==> see my comment at the end of the post, I actually think your paragraph at the end makes a very good case for why sharing this information does make sense, but I'd frame it like that from the beginning: "we just wrote about how lack of human judgement in the AI engineering loop leads to poor agents, here is the living proof of that, even with the newest, smartest models" (also, even though the dataset looked good at first glance, digging deeper manually revealed some shortcomings). Readers will go into the article with a different mindset, which makes more sense imo
